README.md
# Documentation

## Onboarding

Welcome to the team! Please first read the following document to gain a good understanding about our trading system.

### Basic knowledge

- About basic finance terminologies, investopedia is your bestfriend. Read the definition on investopedia and do a brief research before asking.
- Watch the following video from MIT opencourseware https://ocw.mit.edu/courses/mathematics/18-s096-topics-in-mathematics-with-applications-in-finance-fall-2013/video-lectures/lecture-1-introduction-financial-terms-and-concepts/
- About mathematics, newcomers should have basic knowledge in calculus, statistics and probability. In the future, we may work on stochastic calculus and differential equations. Experience in math modeling is recommended.
- Specific documents will be given to people working on specific areas. Contact Hoang and Long for more details.

### What we do
As we are yet to have an official name, we will refer to this system as "tradingsystem". Tradingsystem is basically an algorithmic trading system, but we don't limit ourselves to quant as we also do technical analysis and fundamentals analysis. At the moment, we are trading forex and we may expand to stocks and derivatives in the future.  

Algorithmic trading is an extremely rewarding academic-wise. We have several projects in software engineering, economics, finance mathematics and machine learning for everyone to work on. Here, we have a list of open projects https://docs.google.com/spreadsheets/d/1BJZxA-PRkMm2HwMaKohUK1VaKiOfIIbV8SNgL5QiMrU/edit#gid=0

## Infra Description: 
Our system mostly uses Python 3. It is advised for newcomers to have good knowledge in Python and feel comfortable in using GitHub.

### Authorization:
settings.py + pyalgo.cfg: to access trading account. 

Two main threads:
1. streaming.py: output - real time EURUSD price from Oanda as TICK events (check event.py) 
2. strategy.py: analyze TICK events to determine when to enter. output - SIGNAL events.

### Strategies:
See strategy folder. There are three main strategies:  
- Mean-reversion
- Momentum
- Moving averages

### Risk management:
1. portfolio.py: receive SIGNAL event - conduct risk management. Output - ORDER event. (under dev)

### Simulation - Backtesting:
Expectation: modify strategy.py --> strategy_session.py. User could choose to backtest or trade live. 
Would import BacktestBase.py (under dev): output: statistical indicators of strategy's performance. 
new_hist.py: download historical data for analysis and building backtests. 

### Evaluation:
statisticaltest.py: output - statistical analysis to evaluate strategy's performance. 

Note on Oanda API: http://developer.oanda.com/rest-live-v20/introduction/
- Must connect using Context method. i.e ctx.Connect(args)


## Dependencies
Python 3.5.12 (therefore remember to install Python 3 dependencies)

### API used:
### Core APIs
- numpy
```
pip3 install numpy
```
- pandas
```
pip3 install pandas
```
- scipy
```
pip3 install scipy
```
- queue
```
pip3 install queuelib
```
- time

### Broker APIs
- Oanda
```
pip3 install Oanda
```
### Risk managing and strategy developing APIs
- portfolioopt (has a variety of risk management tools)
```
pip3 install portfolioopt
```
- statsmodels (for statistics indicators)
```
pip3 install statsmodels
```
- johansen (Johansen test)
```
pip3 install johansen
```

### Other
- threading
- seaborn (for data visualization)

## Other
Contact long.le@minerva.kgi.edu and hoang.nguyen@minerva.kgi.edu for more details. Task distribution will be according to project list google sheets file.


changelog.txt
Log

04/04/17
*trading/hist.py
- Fixing date time to RFC3339 format
- Fixing download part

Created new_hist.py

Problems:
- Need to reformat data (solved)
- Try to export to hdf5 format (solved)

05/04/17
Moved new_hist.py to main folder

*new_hist.py
- Added a mock strategy

Problems:
- Strategy not working (solved)

07/04/17
*new_hist.py
- Showing plot
- Strategy checked and worked

Problems:
- Add a y = 1 line for plot (solved using grid)
- Need to store data (solved - hdf5)
- Create a strategy class (solved 09/04/17)
- Create a for loop for repetitive data chunk download (Solved)
	- Currently having issues with prices somehow being empty
	DataFrame(solved - dataframe actually worked as shown)
	- A few trials show that there is nothing wrong with data retrieval, but rather appending dataframe doesn't work
	- Solved, check here for solution: http://stackoverflow.com/questions/16597265/appending-to-an-empty-data-frame-in-pandas

Homework (Hoang):
- Read "Master databases for algorithmic trading" by quantstart
- Read chapter 3 of "Python for Algorithmic Trading"

09/04/17
added momentum strategy to replace test random strat.

Problems:
- Need a method to track positions and P&L
- Need experiments with various momentum timeframes.
- Crude fix for local var problem. Leaner code might be possible. 

#TODO: scale up to run multiple strategies
#TODO: multiple currency pairs

10/04/17
Some discussions on financial data:
- Considering AWS or DigitalOcean for data storing
- Already had steaming.py, so don't have to work on it
- Might have to consider practical optimization regarding discrepancies between original price and bid/ask price

Begin working on strategies
Reading assignment (Hoang):
- https://en.wikipedia.org/wiki/Mathematical_finance
- Algorithmic Trading Winning Strategies



=======
add position management: position.py
add placeholder for portfolio.py

Problems:
- Bug detected: cannot fill order. Possible reason: no margin left. 
- Need to check with new API for possible changes, particularly "SHORT" side
#TODO: fill NaN positions (seems not to work - need revised)

16/04/2017
Wrote a self-reflection about the CME challenge, also propose a working mechanism for the summer interns.

17/04/2017
Working on:
- Update data visualization for new_hist.py using seaborn and potentially bokeh.
- Writing mean-reversion strategy. First, we need to figure out how to feed the data into the system.

*On data visualization:
Some tutorials to look into
https://www.oreilly.com/learning/data-visualization-with-seaborn
http://pbpython.com/visualization-tools-1.html

Also, besides seaborn, there bokeh and matplotlib as potential libraries. Some evaluations:
- Matplotlib is for basic plotting -- bars, pies, lines, scatter plots, etc.

- Seaborn is for statistical visualization -- use it if you're creating heatmaps or somehow summarizing your data and still want to show the distribution of your data

- Bokeh is for interactive visualization -- if your data is so complex (or you haven't yet found the "message" in your data), then use Bokeh to create interactive visualizations that will allow your viewers to explore the data themselves

*new_hist.py
Trying out seaborn api for a better data visualization.
Stuff to try:
- COLH dataviz
- Time-series walk
- Growth visualization


*mean-reversion.py
First trying to understand the structure, it involves event.py, streaming.py and trading.py

22/04/2017
Resumed working
*mean-reversion.py
Implementing Hurst exponent algorithm


23/04/2017
Done with Hurst, created a Hurst.py so that in the future we can use Hurst test for different strategies
Some helpful links:
https://www.quantopian.com/posts/hurst-exponent
https://www.quantopian.com/posts/some-code-from-ernie-chans-new-book-implemented-in-python

24/04/2017
Reorganize GitHub into folders for easy future reference. 
#TODO: 
- change import in major .py programs to reflect new changes. 
- create a readme for statistical tests for instructions.

25/04/2017
*statisticaltest.py
Implementing Johansen test, ADF and CADF test

27/04/2017
Implement portfolio management
Problem:
- BEWARE: Not tested with all cases. UNSTABLE. But provide a good framework
regardless for future strategies. 

28/04/2017
update position.py using Decimal

06/05/2017
turn new_hist.py into a module for implementation in BacktestBase.py 
new change: mv new_hist.py price_retrieval.py
Problem:
- Not tested. Need debug. 
04/05/2017
Finished Johansen

Remember to install johansen and stock API

05/05/2017
Cleaned up new_hist.py, vectorized file will import historical data from new_hist.py
Working on CAPM test and vectorized backtesting

06/05/2017
turn new_hist.py into a module for implementation in BacktestBase.py
new change: mv new_hist.py price_retrieval.py
Problem:
- Not tested. Need debug.

08/05/2017
*VectMomentum.py
- Not finished, needs parameterizing and change yahoo data with Oanda data
- Needs to find a way to export to files or plots
- The code is extremely slow, it takes 11 seconds for 5 tcost data points, 38 seconds for 24 data points
To do:
- Plot the results

10/05/2017
*momentum.py
- Redirect to new event directory
- Parameterize momentum

*VectMomentum.py
- Remove comments
- Add plots

11/05/2017
*VectMomentum.py
- Calls Hurst and MDD to test time series performance
- Add Forex symbol list
https://finance.yahoo.com/connection/currency-converter-cache?date=20150307

*statisticaltest.py
- Add MDD indicator

15/05/2017
*VectMomentum.py
- Change from Yahoo finance data to Oanda for intraday data
Current problem: 
- Oanda data identify itself as object, not DataFrame
- Need to reformat time

*statisticaltest.py
- Adds MDD

*hist.py
- Group the code into methods

20/05/2017
*VectMomentum.py
- Change from Yahoo finance data to Oanda for intraday data

*statisticaltest.py
- Adds MMI (Market Meanness Index)

24/05/2017
*statisticaltest.py
- Fixing Oanda data

27/05/2017
Added weighted moving average strategy

30/05/2017
*VectMomentum.py

new_hist.py
import pandas as pd

import datetime as dt
import v20

import configparser

import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pylab

# new_hist.py

config = configparser.ConfigParser()
config.read('pyalgo.cfg')
access_token = config['oanda_v20']['access_token']

# v20.context
ctx = v20.Context(
    'api-fxpractice.oanda.com',
    443,
    True,
    application='sample_code',
    token=config['oanda_v20']['access_token'],
    datetime_format='RFC3339'
)

response = ctx.account.instruments(config['oanda_v20']['account_id'])
r = response.get('instruments')


def export():
    # dateTime formatting
    suffix = '.000000000Z'
    time1 = dt.datetime(2016, 8, 1, 0, 0, 0)
    d1 = time1.isoformat('T') + suffix
    time2 = dt.datetime(2016, 8, 2, 0, 0, 0)
    d2 = time2.isoformat('T') + suffix
    time_unit = dt.timedelta(1)
    limit = dt.datetime(2016, 8, 15, 0, 0, 0)
    limit = limit.isoformat('T') + suffix

    # data chunking
    prices = pd.DataFrame()
    dates = pd.date_range(start=d1, end=limit, freq='D')
    for i in range(len(dates) - 1):
        d1 = str(dates[i]).replace(' ', 'T')
        d2 = str(dates[i + 1]).replace(' ', 'T')
        candle = ctx.instrument.candles(
            instrument='EUR_USD',
            fromTime=d1,
            toTime=d2,
            granularity='M1',
            price='A'
        )
        data = candle.get('candles')
        data = [cs.dict() for cs in data]  # turn data into dict, pretty important
        for cs in data:
            cs.update(cs['ask'])
            del cs['ask']
        Kappa = pd.DataFrame(data)
        prices = prices.append(Kappa)
    return prices


# translate the data into dictionary and pandas DataFrame

# create dataframe
# prices = pd.DataFrame(data)

def showdata():
    price = export()
    print(price)
    price["time"] = pd.to_datetime(price["time"])
    price = price.set_index("time")
    price.index = pd.DatetimeIndex(price.index)
    price.to_hdf("data.h5", "data", format="table")


# vectorized backtest momentum strategy
"""
prices[["c", "l", "h", "o"]] = prices[["c", "l", "h", "o"]].astype("float64")
prices = prices.rename(columns={"c": "closeAsk", "l":"lowAsk",
                            "h": "highAsk", "o":"openAsk"})
prices[["closeAsk", "volume"]].head()

prices["returns"] = np.log(prices["closeAsk"] / prices["closeAsk"].shift(1))

cols = []

for momentum in [5, 15, 30, 60, 120]:
    col = "position_%s" % momentum
    prices[col] = np.sign(prices["returns"].rolling(momentum).mean())
    cols.append(col)

strats = ["returns"]

for col in cols:
    strat = "strategy_%s" % col.split("_")[1]
    prices[strat] = prices[col].shift(1) * prices["returns"]
    strats.append(strat)


prices[strats].dropna().cumsum().apply(np.exp).plot()
plt.grid()
plt.show()
"""
